src/Workspaces/Remote/Core/BrokeredServiceConnection.cs

```diff
@@ -143,26 +146,76 @@ public override async ValueTask<Optional<TResult>> TryInvokeAsync<TResult>(Solut
-            var readerTask = Task.Run(async () => await reader(clientStream, cancellationToken).ConfigureAwait(false), cancellationToken);
-            await Task.WhenAll(writerTask, readerTask).ConfigureAwait(false);
+            // Create new tasks that both start executing, rather than invoking the delegates directly
+            // to make sure both invocation and reader start executing and transfering data.
+
+            var writerTask = Task.Run(async () =>
+            {
+                try
+                {
+                    await invocation(service, pipe.Writer, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    // Ensure that the writer is complete if an exception is thrown
+                    // before the writer is passed to the RPC proxy. Once it's passed to the proxy 
+                    // the proxy should complete it as soon as the remote side completes it.
+                    await pipe.Writer.CompleteAsync(e).ConfigureAwait(false);
+
+                    throw;
+                }
+            }, cancellationToken);
+
+            var readerTask = Task.Run(
+                async () =>
+                {
+                    Exception? exception = null;
+
+                    try
+                    {
+                        return await reader(pipe.Reader, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception e) when ((exception = e) == null)
+                    {
+                        throw ExceptionUtilities.Unreachable;
+                    }
+                    finally
+                    {
+                        await pipe.Reader.CompleteAsync(exception).ConfigureAwait(false);
+                    }
+                }, cancellationToken);
+
+            try
+            {
+                await Task.WhenAll(writerTask, readerTask).ConfigureAwait(false);
+            }
+            catch (AggregateException aggregateException) when (cancellationToken.IsCancellationRequested)
+            {
+                aggregateException.Handle(e => e is OperationCanceledException);
+                cancellationToken.ThrowIfCancellationRequested();
+            }
 
             return readerTask.Result;
         }
 
         private bool ReportUnexpectedException(Exception exception, CancellationToken cancellationToken)
         {
+            if (exception is OperationCanceledException)
+            {
+                // It's a bug for a service to throw OCE based on a different cancellation token than it has received in the call.
+                // The server side filter will report NFW in such scenario, so that the underlying issue can be fixed.
+                // Do not treat this as a critical failure of the service for now and only fail in debug build.
+                Debug.Assert(cancellationToken.IsCancellationRequested);
```

Comment:
I'm not sure this comment is exactly what we mean. It's generally acceptable to throw a cancellation exception with a different cancellation token provided the original input token _also_ has cancellation requested.

Reply:
Right. The comment doesn't mention the token associated with the exception, but the token that triggered OCE. The token received by the call needs to be signaled when the exception is thrown, otherwise it's a bug.